pkg/metrics/collector/sockstat.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2018 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type sockStatCollector struct {
+	socketInuseDesc  *prometheus.Desc
+	socketTwDesc     *prometheus.Desc
+	socketMemoryDesc *prometheus.Desc
+}
+
+var pageSize = os.Getpagesize()
+
+func init() {
+	registerCollector("socket", NewSockStatCollector)
+}
+
+// NewSockStatCollector returns a new Collector exposing socket stats.
+func NewSockStatCollector() (Collector, error) {
+	return &sockStatCollector{
+		socketInuseDesc: prometheus.NewDesc(
+			"num_inuse_sockets",
+			"Number of inuse sockets",
+			[]string{"protocol"}, nil,
+		),
+		socketTwDesc: prometheus.NewDesc(
+			"num_tw_sockets",
+			"Number of sockets in time wait state",
+			nil, nil,
+		),
+		socketMemoryDesc: prometheus.NewDesc(
+			"socket_memory",
+			"Amount of memory used by sockets",
+			nil, nil,
+		),
+	}, nil
+}
+
+type socketStats struct {
+	tcpInuse uint64 // num of tcp inuse sockets
+	udpInuse uint64 // num of udp inuse sockets
+	tcpTw    uint64 // num of tcp tw sockets
+	mem      uint64 // sum of tcp and udp memory
+}
+
+func getSockStats(fileName string) (*socketStats, error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseSockStats(file)
+}
+
+func parseSockStats(r io.Reader) (*socketStats, error) {
+	var ss socketStats
+	var lineCount int = 0
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		lineCount++
+		in := scanner.Text()
+		line := strings.Split(in, " ")
+		if len(line) < 2 || len(line[0]) < 2 || line[0][len(line[0])-1] != ':' {
+			return nil, fmt.Errorf("invalid input %q", in)
+		}
+		// Remove trailing ':'.
+		protocol := line[0][:len(line[0])-1]
+		if protocol == "TCP" || protocol == "TCP6" {
```

Comment:
let factor this into several functions for easier maintenance:

```go
func (s*socketStats) merge(o *socketStats) {
  s.tcpInUse = o.tcpInUse
  ...
  // Sum memory used across protocols.
  s.mem += o.mem
}

func extractSockStatUint(field string, s string, out interface{}) error {
  v, err := strconv.ParseUint(s, 10, 64)
  if err != nil {
    return nil, fmt.Errorf("invalid value %q for field %q in sockstats", s, field)
  }
  outValue = out.(*uint64)
  *outValue = v
  return nil
}

type sockStatEntry struct {
  out interface{}
  extract func(string, string, interface{}) error
}

func extractSockStatFields(fields []string, entries map[string]sockStatEntry) error {
  for i := 0; i < len(fields) && i+1 < len(fields); i += 2 {
    k, v := fields[i], fields[i+1]
    if e, ok := entries[k]; !ok {
      continue
    }
    if err := e.extract(k, v, e.out); err != nil {
      return err
    } 
  }
  return nil
}

func parseLine(in string) (*socketStats, error) {
  line := strings.Split(in, " ")
  if len(line) < 2 || len(line[0]) < 2 || line[0][len(line[0])-1] != ':' {
    return nil, fmt.Errorf("invalid line %q", in)
  }

  tcpEntries := map[string]*sockStatEntry{
      "inuse": {&ret.tcpInUse, extractSockStatUint},
      "tw": {&ret.tcpTw, extractSockStatUint},
      "mem": {&ret.mem, extractSockStatUint},
  }

  udpEntries := map[string]*sockStatEntry{
    "inuse": {&ret.tcpInUse, extractSockStatUint},
    "tw": {&ret.tcpTw, extractSockStatUint},
  }

  ret := &socketStats{}
  protocol := line[0][:len(line[0])-1]
  rest := line[1:]

  switch protocol {
  case "TCP", "TCP6":
    if err := extractSockStatFields(rest, tcpEntries); err != nil {
      return err
    }
  case "UDP", "UDP6":
    if err := extractSockStatFields(rest, udpEntries); err != nil {
      return err
    }
  }
  return nil
}

func parseSockStats(r io.Reader) (*socketStats, error) {
  ...
  ret := &socketStats{}
  for scanner.Scan() {
    lineCount++
    if s, err := parseLine(scanner.Text()); err == nil {
      ret.merge(s)
    } else {
      return nil, err
    }
  }
  ...
}
```

Reply:
Updated.